### PR TITLE
Butte boost

### PR DESCRIPTION
**Pull request by magnet**
**New features**
- New IR format
- Union support
- RPC support
- Root types
- Required/optional fields and default values (working but may still need some work, it's using the `flatbuffers`'s crate design and we might want to revisit)
- Tables generic over buffer type (needs to impl `AsRef<[u8]>`)

**Bug fixes**
- Empty tables not building
- Broken vector support

**Removed because incomplete/untested**
- Nested flatbuffers
- Includes

### CODE REVIEW

**Comment by lovesegfault** [comment]
Hey @magnet I'm on vacation right now so I'll have time to review this, hopefully this week :)

File: butte-build/src/ir/types.rs
Comment (lovesegfault) [inline butte-build/src/ir/types.rs:235]: Should this be `&'static str`?

File: butte-build/src/ir/types.rs
Comment (lovesegfault) [inline butte-build/src/ir/types.rs:382]: Is it called `Bidi` elsewhere too? If not I'd prefer just writing out `Bidirectional`.

File: butte-build/src/ir/types.rs
Comment (lovesegfault) [inline butte-build/src/ir/types.rs]: I never liked `DottedIdent`. I think we should find a better name. I previously suggested `QualifiedIdent`. What do you think?

**Review by lovesegfault — COMMENTED** [review]
Overall this is a work of art.

I left some small comments I'd like to hear back on before approving, and I think @cpcloud should take a look as well around `ir/transform.rs` and `codegen.rs`

File: butte-build/src/ir/types.rs
Comment (magnet) [reply butte-build/src/ir/types.rs:235]: In practice this doesn't matter much as this is just used for "displaying", e.g in doc we generate (though, we never generate doc using this, as strings can't have defaults).  In general though, the string slice returned from buffers will usually have the (elided) lifetime of the buffer, and that lifetime wouldn't be `'static`.

File: butte-build/src/ir/types.rs
Comment (magnet) [reply butte-build/src/ir/types.rs:382]: It's called `"bidi"` in the flatbuffers RPC spec, and I'd be happy to call it either way, but Rust usually has shorthands (e.g `Arc` or `Vec`) so I found `Bidi` a good match.

File: butte-build/src/ir/types.rs
Comment (magnet) [reply butte-build/src/ir/types.rs]: Agreed, I don't like `DottedIdent` either, and `QualifiedIdent` works for me.  The method that convert simple idents to namespaced ones is called `make_fully_qualified_ident` which would also match this naming better ("fully" referring here to the fact that the namespace is specified).

**Comment by magnet** [comment]
> Overall this is a work of art.

:blush:  Thanks! This still lacks the polish it needs, but I think it's in a usable state for users who don't rely on everything flatbuffers support (structs, non-supported attributes, ...), and in a hackable state for those who do but want to provide a PR in butte. Hopefully with some more effort we could get an alpha version on `crates.io`.

> I left some small comments I'd like to hear back on before approving, and I think @cpcloud should take a look as well around `ir/transform.rs` and `codegen.rs`

Thanks. For reference, here's the PR adding `flatbuffers` support to `tonic` based on this branch: https://github.com/hyperium/tonic/pull/250

File: butte-build/src/ir/types.rs
Comment (lovesegfault) [reply butte-build/src/ir/types.rs:235]: Sounds good

File: butte-build/src/ir/types.rs
Comment (lovesegfault) [reply butte-build/src/ir/types.rs]: Nice, let’s make the change then, unless @cpcloud strongly opposes. I don’t remember his thoughts on this when we last discussed it.

File: butte-build/src/ir/types.rs
Comment (lovesegfault) [reply butte-build/src/ir/types.rs:382]: If it’s `bidi` in the spec let’s leave `Bidi` here :)

File: butte-build/src/ir/types.rs
Comment (cpcloud) [reply butte-build/src/ir/types.rs]: Gut it!

File: butte-build/Cargo.toml
Comment (cpcloud) [inline butte-build/Cargo.toml:50]: I recently discovered this crate and so far am loving it for testing big trees of things.

File: butte-build/src/ir/types.rs
Comment (magnet) [reply butte-build/src/ir/types.rs]: Done!

**Review by lovesegfault — APPROVED** [review]
chef-hands

File: butte-build/src/ast/parser.rs
Comment (cpcloud) [inline butte-build/src/ast/parser.rs:1422]: Excellent. Thanks for finding and fixing.

File: butte/src/primitives.rs
Comment (cpcloud) [inline butte/src/primitives.rs]: When I change this to:

```
        let off = {
            let r = buf.as_ref();
            let slice = r.get(loc..loc + SIZE_UOFFSET).ok_or(Error::OutOfBounds)?;
            read_scalar::<u32>(slice)? as usize
        };
```
I start seeing a Clippy warning:
```
warning: this call to `as_ref` does nothing
   --> butte/src/primitives.rs:177:21
    |
177 |             let r = buf.as_ref();
    |                     ^^^^^^^^^^^^ help: try this: `buf`
    |
    = note: `#[warn(clippy::useless_asref)]` on by default
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_asref
```
I guess then there are a few questions:

1. Is the `as_ref()` call necessary?
1. If so, what effect does it have?
1. And curiously, why does clippy warn in one case but not the other (this definitely doesn't need to be answered to merge this PR)?

File: butte/src/primitives.rs
Comment (cpcloud) [inline butte/src/primitives.rs:196]: Here the `as_ref()` is clearly needed because of the trait bound on `T::Buf` (just noting this for myself).

File: butte/src/table.rs
Comment (cpcloud) [inline butte/src/table.rs:27]: Nice work. This should make `Table` a lot more flexible WRT downstream consumers.

File: butte-build/src/ir/types.rs
Comment (cpcloud) [inline butte-build/src/ir/types.rs:160]: Interesting, I did not know this.

File: butte-build/src/ir/types.rs
Comment (cpcloud) [inline butte-build/src/ir/types.rs:162]: That is also nice.

File: butte-build/src/ir/types.rs
Comment (cpcloud) [inline butte-build/src/ir/types.rs:164]: It would be nice to support them, but I'd say this is low priority right now IMO.

Comment (cpcloud) [reply butte-build/src/ir/types.rs:164]: Also supporting them in the parser is probably not too hard, it's elsewhere that things tend to be more challenging IME :)

File: butte-build/src/ir/types.rs
Comment (cpcloud) [inline butte-build/src/ir/types.rs:304]: Any reason to use `()` as opposed to a proper error type?

File: butte-build/src/ir/types.rs
Comment (cpcloud) [inline butte-build/src/ir/types.rs]: Maybe make this `Self`?

File: butte-build/src/ir/transform.rs
Comment (cpcloud) [inline butte-build/src/ir/transform.rs:45]: nit: line break after this `}`

File: butte-build/src/codegen.rs
Comment (cpcloud) [inline butte-build/src/codegen.rs:104]: Just out of curiosity could we avoid adding this flag if we made `wrap_refs_types` into an `Option<&impl ToTokens>`?

Comment (cpcloud) [reply butte-build/src/codegen.rs:104]: I'd be happy to do this in a follow up in the interest of merging ASAP.

File: butte-build/src/ir/transform.rs
Comment (cpcloud) [inline butte-build/src/ir/transform.rs:385]: Formatting here looks a bit off. Should the two previous lines be indented?

File: butte-build/src/ir/transform.rs
Comment (cpcloud) [inline butte-build/src/ir/transform.rs:374]: Should we parse these into enums instead of converting them during codegen? Happy to do this in a follow up.

File: butte-build/src/codegen.rs
Comment (cpcloud) [inline butte-build/src/codegen.rs]: nit: kill this extra newline.

File: butte-build/src/codegen.rs
Comment (cpcloud) [inline butte-build/src/codegen.rs:281]: Nice, thanks for adding `(required)` support!

File: butte-build/src/codegen.rs
Comment (cpcloud) [inline butte-build/src/codegen.rs]: nit: can we move the first comma inside the parens to avoid writing two commas?

File: butte-build/src/codegen.rs
Comment (cpcloud) [inline butte-build/src/codegen.rs]: nit: similar comment to the above about two commas.

File: butte-build/src/codegen.rs
Comment (cpcloud) [inline butte-build/src/codegen.rs:939]: :grimacing:

**Review by cpcloud — APPROVED** [review]
The verdict: agree with @lovesegfault, this is wonderful. Thanks for doing this. I have a few minor questions about follow-ups as well as style fixes, but I'm extremely happy with this PR. :tada:

File: butte/src/primitives.rs
Comment (magnet) [reply butte/src/primitives.rs]: This is a remainder of a larger code change I had made where the buffer was generic here as well, I'll look into it.

File: butte-build/src/ir/types.rs
Comment (magnet) [reply butte-build/src/ir/types.rs:304]: Thanks for spotting this, it was just faster during dev time.

File: butte-build/src/ir/types.rs
Comment (magnet) [reply butte-build/src/ir/types.rs]: I'm not entirely convinced `Self` is always more readable, though it does save some keystrokes for complex types, e.g if it was `Self` already, with those 2 lines I wouldn't see what type this is about :).  Having `Self` and `QualifiedIdent` in the same method gives more context IMO, though I don't have strong opinions with this and will change if you want this for consistency.

File: butte-build/src/ir/transform.rs
Comment (magnet) [reply butte-build/src/ir/transform.rs:45]: will do

File: butte-build/src/codegen.rs
Comment (magnet) [reply butte-build/src/codegen.rs:104]: > Just out of curiosity could we avoid adding this flag if we made wrap_refs_types into an Option<&impl ToTokens>?

We couldn't, this helps with complex types like `Vector` or `WIPOffset` that are not wrapped the same as the "internal" type.  I'm not really happy with this logic in general, but first step is to make it work, then make it nice :)

File: butte-build/src/ir/transform.rs
Comment (magnet) [reply butte-build/src/ir/transform.rs:385]: I'll have a look, but this is likely `rustfmt`'s doing.

File: butte-build/src/ir/transform.rs
Comment (magnet) [reply butte-build/src/ir/transform.rs:374]: I think it's nice for the parsing layer to remain dumb, and for the semantic layer to pick the supported attributes into enums.

File: butte-build/src/codegen.rs
Comment (magnet) [reply butte-build/src/codegen.rs:939]: We can't do much about that one right, since we generate pretty complex types with like `Option<WIPOffset<Vector<'a, SomeTable<...>>`.  With this though, the generated modules pass Clippy's default lints.

Comment (cpcloud) [reply butte-build/src/codegen.rs:939]: Yeah, totally :)

File: butte-build/src/ir/types.rs
Comment (cpcloud) [reply butte-build/src/ir/types.rs:304]: Roger that.

File: butte/src/primitives.rs
Comment (magnet) [reply butte/src/primitives.rs]: Fixed

File: butte-build/src/ir/types.rs
Comment (cpcloud) [reply butte-build/src/ir/types.rs]: Fair enough, I'll leave this up to you. I find `Self` more readable in `impl`s that are smaller, FWIW.

File: butte-build/src/codegen.rs
Comment (cpcloud) [reply butte-build/src/codegen.rs:104]: Absoultely, carry on.

File: butte-build/src/ir/transform.rs
Comment (cpcloud) [reply butte-build/src/ir/transform.rs:385]: Cool, I wouldn't worry about it too much.

File: butte-build/src/ir/transform.rs
Comment (cpcloud) [reply butte-build/src/ir/transform.rs:374]: Yeah, I go back and forth on just _how_ dumb it should be (agree that it should be dumb). In this case it seems like a little more intelligence would be okay, but I also don't feel that strongly about it. I may tinker around a bit after this is merged.

File: butte-build/src/codegen.rs
Comment (magnet) [reply butte-build/src/codegen.rs]: Fixed in a later commit

File: butte-build/src/codegen.rs
Comment (magnet) [reply butte-build/src/codegen.rs]: Done, thanks for the suggestion

File: butte-build/src/codegen.rs
Comment (magnet) [reply butte-build/src/codegen.rs]: Fixed

**Pull request merged**
